### PR TITLE
Let it snow

### PR DESCRIPTION
**Pull request by trontrytel**
### Description

This adds snow to Atmos interface

I still need to:
- [x] check results from longer simulation runs (make squall line run longer)
- [x] update the docs about coupling to total energy
- [x] rebase, format, probably add some API docstrings

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by valeriabarra** [comment]
I just want to say that I love the PR's name. It made me smile!

**Comment by charleskawczynski** [comment]
Such a festive PR ❄️ 😄

**Review by charleskawczynski — COMMENTED** [review]
At a first glance, this is looking pretty good from my end

**Comment by trontrytel** [comment]
> Such a festive PR ❄️ 😄

I heard that parts of Massachusetts are on winter storm watch. Maybe soon we will be too!

**Comment by djarecka** [comment]
Before and after merging ;-)

**Comment by trontrytel** [comment]
150 minutes of squall line. All units are in [g/kg]

test4

**Comment by trontrytel** [comment]
@charleskawczynski  - It's ready for review. I'll add it to microphysics qa in the next PR. This is already too long and I don't want to delay even more.

@slifer50 - you were right about the signs in the previous iteration of this PR. I managed to confuse myself with the source/sink terminology in the docs. Thanks for catching that. I added documentation on coupling to state variables. I hope it clears it a little, though coupling to energy is a little complicated.

File: docs/src/Theory/Atmos/Microphysics.md [hunk 1]
```diff
@@ -678,32 +681,35 @@ If ``T > T_{freeze}``:
...
+  the ``\mathcal{S}_{q_{rai}}`` source has to be multiplied by ``\rho``
+  before adding it as one of the source terms to ``\rho q_{rai}`` state variable.
 For the conservation equation for
 [total energy](https://clima.github.io/ClimateMachine.jl/latest/Theory/Atmos/AtmosModel/#Energy),
   the sink due to removing ``q_{tot}`` is computed as:
 ```math
 \begin{equation}
-\left. \sum_{j\in\{v,l,i\}}(I_j + \Phi)  \rho C(q_j \rightarrow q_p) \right|_{precip} =
+\left. \mathcal{S}_{\rho e}} \right|_{precip} =
+  \left. \sum_{j\in\{v,l,i\}}(I_j + \Phi)  \rho C(q_j \rightarrow q_p) \right|_{precip} =
```
Comment (charleskawczynski) [inline docs/src/Theory/Atmos/Microphysics.md:712]: Looks like this broke the latex equation

File: docs/src/Theory/Atmos/Microphysics.md [hunk 2]
```diff
@@ -714,6 +720,115 @@ where:
...
+The contribution from
+  cloud liquid water to rain autoconversion,
+  cloud liquid water accretion by rain,
+  and rain evaporation
+  is computed as
+```math
+\begin{equation}
+\left. \mathcal{S}_{\rho e}} \right|_{precip} =
+  - (I_l + \Phi) \rho \, \mathcal{S}_{q_{rai}}
+\end{equation}
```
Comment (charleskawczynski) [inline docs/src/Theory/Atmos/Microphysics.md:771]: This equation is broken also

File: src/Atmos/Model/atmos_tendencies.jl [hunk 3]
```diff
@@ -26,16 +28,39 @@ filter_source(
...
+filter_source(
+    pv::PV,
+    ::MoistureModel,
+    s::RainSnow_1M{PV},
+) where {PV <: LiquidMoisture} = nothing
+filter_source(
+    pv::PV,
+    ::NonEquilMoist,
+    s::RainSnow_1M{PV},
+) where {PV <: IceMoisture} = s
```
Comment (charleskawczynski) [inline src/Atmos/Model/atmos_tendencies.jl:56]:
```suggestion
filter_source(
    pv::PV,
    ::NonEquilMoist,
    s::RainSnow_1M{PV},
) where {PV <: Union{LiquidMoisture, IceMoisture}} = s
filter_source(
    pv::PV,
    ::MoistureModel,
    s::RainSnow_1M{PV},
) where {PV <: LiquidMoisture} = nothing
```

File: src/Atmos/Model/multiphysics_types.jl [hunk 4]
```diff
@@ -119,3 +121,216 @@ function compute_rain_params(atmos, state, aux, t, ts)
...
+
+RainSnow_1M() = (
+    RainSnow_1M{Mass}(),
+    RainSnow_1M{Energy}(),
+    RainSnow_1M{TotalMoisture}(),
+    RainSnow_1M{LiquidMoisture}(),
+    RainSnow_1M{IceMoisture}(),
+    RainSnow_1M{Rain}(),
+    RainSnow_1M{Snow}(),
+)
```
Comment (charleskawczynski) [inline src/Atmos/Model/multiphysics_types.jl:143]: ❤️

File: src/Atmos/Model/multiphysics_types.jl [hunk 5]
```diff
@@ -119,3 +121,216 @@ function compute_rain_params(atmos, state, aux, t, ts)
...
+
+export RainSnow_1M
+"""
+    RainSnow_1M{FT} <: AbstractSource
+A collection of source/sink terms related to 1-moment rain and snow microphysics
+"""
+struct RainSnow_1M{
+    PV <:
+    Union{Mass, Energy, TotalMoisture, LiquidMoisture, IceMoisture, Rain, Snow},
+} <: TendencyDef{Source, PV} end
```
Comment (charleskawczynski) [inline src/Atmos/Model/multiphysics_types.jl:133]:
```suggestion
struct RainSnow_1M{
    PV <:
    Union{Mass, Energy, Moisture, Rain, Snow},
} <: TendencyDef{Source, PV} end
```

File: src/Atmos/Model/precipitation.jl [hunk 6]
```diff
@@ -100,39 +147,104 @@ function compute_gradient_flux!(
...
+    t::Real,
+    D_t,
+)
+    d_q_rai = (-D_t) .* diffusive.precipitation.∇q_rai
+
+    flux_second_order!(precip, flux, state, d_q_rai)
+end
+function flux_second_order!(precip::RainModel, flux::Grad, state::Vars, d_q_rai)
+    flux.precipitation.ρq_rai += d_q_rai * state.ρ
+end
```
Comment (charleskawczynski) [inline src/Atmos/Model/precipitation.jl:183]: We could use the new tendency spec for this, but I'm fine leaving that for a separate PR.

**Review by charleskawczynski — COMMENTED** [review]
I left some comments about a few broken latex equations (there are a few more). Overall looks fine from my angle on the code side. Let it snow! ❄️

File: docs/src/Theory/Atmos/Microphysics.md (hunk 1 above)
Comment (trontrytel) [reply docs/src/Theory/Atmos/Microphysics.md:712]: damn. I remember adding this line thinking "not gonna compile it, its only one line, what could possibly go wrong". Well apparently everything can. I hope removing one random "}" fixes it.

File: docs/src/Theory/Atmos/Microphysics.md (hunk 2 above)
Comment (trontrytel) [reply docs/src/Theory/Atmos/Microphysics.md:771]: I hope I fixed all of them. Lets see in the next preview

File: src/Atmos/Model/multiphysics_types.jl (hunk 4 above)
Comment (trontrytel) [reply src/Atmos/Model/multiphysics_types.jl:143]: It looks good. I think this source term change improves readability. Go refactor!

File: src/Atmos/Model/precipitation.jl (hunk 6 above)
Comment (trontrytel) [reply src/Atmos/Model/precipitation.jl:183]: Sounds good. Lets do it in the new one. This is already too big PR if you ask me. Who wrote it? ;)

**Comment by trontrytel** [comment]
Seems like the equations are properly rendered now @charleskawczynski

**Review by charleskawczynski — APPROVED** [review]
Looking good!

**Comment by charleskawczynski** [comment]
🤦🏻 That failure was my fault. Feel free to un-split `filter_source` (back to how you previously had it) for now.

**Pull request merged**